Connections leakage occurred if Scaling data consistency check keep failing

## Bug Report

According to commit logs, this issue affects ShardingSphere >= 5.1.0.

This issue happened after #17806 occurred. Let it keep failing and just wait for tens of minutes, then this issue can be reproduced.

### Which version of ShardingSphere did you use?

master - https://github.com/apache/shardingsphere/tree/a393962f5b3eec2d154ad1c2c5d15a758efc499b

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?

ShardingSphere-Proxy

### Expected behavior

Connections are enough.

### Actual behavior

There are hundreds of `HikariDataSource` instances in the Process.

![image](https://user-images.githubusercontent.com/20503072/169272974-c57de140-c2e9-4dea-b635-cccf58ffaca3.png)

```
[INFO ] 2022-05-19 18:21:10.032 [_finished_check_Worker-1] o.a.s.d.p.core.job.FinishedCheckJob - scaling job 0130317c30317c3054317c626d73716c5f7368617264696e67 almost finished.
[INFO ] 2022-05-19 18:21:10.032 [_finished_check_Worker-1] o.a.s.d.p.s.r.s.DefaultSourceWritingStopLock - lock, databaseName=bmsql_sharding, jobId=0130317c30317c3054317c626d73716c5f7368617264696e67
[INFO ] 2022-05-19 18:21:10.042 [_finished_check_Worker-1] o.a.s.d.p.c.a.i.RuleAlteredJobAPIImpl - stopClusterWriteDB, tryLockSuccess=true
[INFO ] 2022-05-19 18:21:10.043 [_finished_check_Worker-1] o.a.s.d.p.core.job.FinishedCheckJob - dataConsistencyCheck for job 0130317c30317c3054317c626d73716c5f7368617264696e67
[INFO ] 2022-05-19 18:21:10.045 [_finished_check_Worker-1] o.a.s.d.p.a.c.TableNameSchemaNameMapping - mapping={pg_trigger=pg_catalog, bmsql_district=public, bmsql_oorder=public, bmsql_stock=public, columns=information_schema, pg_class=pg_catalog, pg_tablespace=pg_catalog, pg_database=pg_catalog, bmsql_customer=public, bmsql_item=public, tables=information_schema, bmsql_new_order=public, bmsql_history=public, bmsql_warehouse=public, pg_inherits=pg_catalog, bmsql_order_line=public, views=information_schema, bmsql_config=public}
[ERROR] 2022-05-19 18:21:11.069 [_finished_check_Worker-1] com.zaxxer.hikari.pool.HikariPool - HikariPool-610 - Exception during pool initialization.
org.postgresql.util.PSQLException: FATAL: sorry, too many clients already
	at org.postgresql.core.v3.ConnectionFactoryImpl.doAuthentication(ConnectionFactoryImpl.java:646)
	at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:180)
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:235)
	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:223)
	at org.postgresql.Driver.makeConnection(Driver.java:400)
	at org.postgresql.Driver.connect(Driver.java:259)
	at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:138)
	at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:354)
	at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:202)
	at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:473)
	at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:554)
	at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:115)
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:112)
	at org.apache.shardingsphere.infra.database.type.DatabaseTypeEngine.getDatabaseType(DatabaseTypeEngine.java:98)
	at org.apache.shardingsphere.infra.database.type.DatabaseTypeEngine.getDatabaseType(DatabaseTypeEngine.java:90)
	at org.apache.shardingsphere.infra.database.type.DatabaseTypeEngine.getFrontendDatabaseType(DatabaseTypeEngine.java:57)
	at org.apache.shardingsphere.mode.metadata.MetaDataContextsBuilder.build(MetaDataContextsBuilder.java:58)
	at org.apache.shardingsphere.mode.manager.memory.MemoryContextManagerBuilder.build(MemoryContextManagerBuilder.java:49)
	at org.apache.shardingsphere.driver.jdbc.core.datasource.ShardingSphereDataSource.createContextManager(ShardingSphereDataSource.java:80)
	at org.apache.shardingsphere.driver.jdbc.core.datasource.ShardingSphereDataSource.<init>(ShardingSphereDataSource.java:64)
	at org.apache.shardingsphere.driver.api.ShardingSphereDataSourceFactory.createDataSource(ShardingSphereDataSourceFactory.java:77)
	at org.apache.shardingsphere.driver.api.ShardingSphereDataSourceFactory.createDataSource(ShardingSphereDataSourceFactory.java:138)
	at org.apache.shardingsphere.driver.data.pipeline.datasource.creator.ShardingSpherePipelineDataSourceCreator.createPipelineDataSource(ShardingSpherePipelineDataSourceCreator.java:43)
	at org.apache.shardingsphere.data.pipeline.core.datasource.PipelineDataSourceFactory.newInstance(PipelineDataSourceFactory.java:44)
	at org.apache.shardingsphere.data.pipeline.core.check.consistency.DataConsistencyChecker.checkCount(DataConsistencyChecker.java:111)
	at org.apache.shardingsphere.data.pipeline.core.check.consistency.DataConsistencyChecker.check(DataConsistencyChecker.java:93)
	at org.apache.shardingsphere.data.pipeline.core.api.impl.RuleAlteredJobAPIImpl.dataConsistencyCheck(RuleAlteredJobAPIImpl.java:288)
	at org.apache.shardingsphere.data.pipeline.core.api.impl.RuleAlteredJobAPIImpl.dataConsistencyCheck(RuleAlteredJobAPIImpl.java:274)
	at org.apache.shardingsphere.data.pipeline.core.job.FinishedCheckJob.dataConsistencyCheck(FinishedCheckJob.java:124)
	at org.apache.shardingsphere.data.pipeline.core.job.FinishedCheckJob.execute(FinishedCheckJob.java:89)
	at org.apache.shardingsphere.elasticjob.simple.executor.SimpleJobExecutor.process(SimpleJobExecutor.java:33)
	at org.apache.shardingsphere.elasticjob.simple.executor.SimpleJobExecutor.process(SimpleJobExecutor.java:29)
	at org.apache.shardingsphere.elasticjob.executor.ElasticJobExecutor.process(ElasticJobExecutor.java:172)
	at org.apache.shardingsphere.elasticjob.executor.ElasticJobExecutor.process(ElasticJobExecutor.java:141)
	at org.apache.shardingsphere.elasticjob.executor.ElasticJobExecutor.execute(ElasticJobExecutor.java:124)
	at org.apache.shardingsphere.elasticjob.executor.ElasticJobExecutor.execute(ElasticJobExecutor.java:100)
	at org.apache.shardingsphere.elasticjob.lite.internal.schedule.LiteJob.execute(LiteJob.java:35)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
[INFO ] 2022-05-19 18:21:11.069 [_finished_check_Worker-1] o.a.s.d.p.s.r.s.DefaultSourceWritingStopLock - releaseLock, databaseName=bmsql_sharding, jobId=0130317c30317c3054317c626d73716c5f7368617264696e67
[INFO ] 2022-05-19 18:21:11.069 [_finished_check_Worker-1] o.a.s.d.p.c.a.i.RuleAlteredJobAPIImpl - restoreClusterWriteDB, before releaseLock, databaseName=bmsql_sharding, jobId=0130317c30317c3054317c626d73716c5f7368617264696e67
[ERROR] 2022-05-19 18:21:11.073 [_finished_check_Worker-1] o.a.s.d.p.core.job.FinishedCheckJob - scaling job 0130317c30317c3054317c626d73716c5f7368617264696e67 finish check failed!
java.lang.IllegalArgumentException: FATAL: sorry, too many clients already
	at org.apache.shardingsphere.infra.database.type.DatabaseTypeEngine.getDatabaseType(DatabaseTypeEngine.java:101)
	at org.apache.shardingsphere.infra.database.type.DatabaseTypeEngine.getDatabaseType(DatabaseTypeEngine.java:90)
	at org.apache.shardingsphere.infra.database.type.DatabaseTypeEngine.getFrontendDatabaseType(DatabaseTypeEngine.java:57)
	at org.apache.shardingsphere.mode.metadata.MetaDataContextsBuilder.build(MetaDataContextsBuilder.java:58)
	at org.apache.shardingsphere.mode.manager.memory.MemoryContextManagerBuilder.build(MemoryContextManagerBuilder.java:49)
	at org.apache.shardingsphere.driver.jdbc.core.datasource.ShardingSphereDataSource.createContextManager(ShardingSphereDataSource.java:80)
	at org.apache.shardingsphere.driver.jdbc.core.datasource.ShardingSphereDataSource.<init>(ShardingSphereDataSource.java:64)
	at org.apache.shardingsphere.driver.api.ShardingSphereDataSourceFactory.createDataSource(ShardingSphereDataSourceFactory.java:77)
	at org.apache.shardingsphere.driver.api.ShardingSphereDataSourceFactory.createDataSource(ShardingSphereDataSourceFactory.java:138)
	at org.apache.shardingsphere.driver.data.pipeline.datasource.creator.ShardingSpherePipelineDataSourceCreator.createPipelineDataSource(ShardingSpherePipelineDataSourceCreator.java:43)
	at org.apache.shardingsphere.data.pipeline.core.datasource.PipelineDataSourceFactory.newInstance(PipelineDataSourceFactory.java:44)
	at org.apache.shardingsphere.data.pipeline.core.check.consistency.DataConsistencyChecker.checkCount(DataConsistencyChecker.java:111)
	at org.apache.shardingsphere.data.pipeline.core.check.consistency.DataConsistencyChecker.check(DataConsistencyChecker.java:93)
	at org.apache.shardingsphere.data.pipeline.core.api.impl.RuleAlteredJobAPIImpl.dataConsistencyCheck(RuleAlteredJobAPIImpl.java:288)
	at org.apache.shardingsphere.data.pipeline.core.api.impl.RuleAlteredJobAPIImpl.dataConsistencyCheck(RuleAlteredJobAPIImpl.java:274)
	at org.apache.shardingsphere.data.pipeline.core.job.FinishedCheckJob.dataConsistencyCheck(FinishedCheckJob.java:124)
	at org.apache.shardingsphere.data.pipeline.core.job.FinishedCheckJob.execute(FinishedCheckJob.java:89)
	at org.apache.shardingsphere.elasticjob.simple.executor.SimpleJobExecutor.process(SimpleJobExecutor.java:33)
	at org.apache.shardingsphere.elasticjob.simple.executor.SimpleJobExecutor.process(SimpleJobExecutor.java:29)
	at org.apache.shardingsphere.elasticjob.executor.ElasticJobExecutor.process(ElasticJobExecutor.java:172)
	at org.apache.shardingsphere.elasticjob.executor.ElasticJobExecutor.process(ElasticJobExecutor.java:141)
	at org.apache.shardingsphere.elasticjob.executor.ElasticJobExecutor.execute(ElasticJobExecutor.java:124)
	at org.apache.shardingsphere.elasticjob.executor.ElasticJobExecutor.execute(ElasticJobExecutor.java:100)
	at org.apache.shardingsphere.elasticjob.lite.internal.schedule.LiteJob.execute(LiteJob.java:35)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: org.postgresql.util.PSQLException: FATAL: sorry, too many clients already
	at org.postgresql.core.v3.ConnectionFactoryImpl.doAuthentication(ConnectionFactoryImpl.java:646)
	at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:180)
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:235)
	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:223)
	at org.postgresql.Driver.makeConnection(Driver.java:400)
	at org.postgresql.Driver.connect(Driver.java:259)
	at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:138)
	at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:354)
	at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:202)
	at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:473)
	at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:554)
	at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:115)
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:112)
	at org.apache.shardingsphere.infra.database.type.DatabaseTypeEngine.getDatabaseType(DatabaseTypeEngine.java:98)
	... 25 common frames omitted
```

--- a/shardingsphere-jdbc/shardingsphere-jdbc-core/src/main/java/org/apache/shardingsphere/driver/data/pipeline/datasource/creator/ShardingSpherePipelineDataSourceCreator.java
+++ b/shardingsphere-jdbc/shardingsphere-jdbc-core/src/main/java/org/apache/shardingsphere/driver/data/pipeline/datasource/creator/ShardingSpherePipelineDataSourceCreator.java
@@ -45,8 +45,7 @@ public final class ShardingSpherePipelineDataSourceCreator implements PipelineDa
         enableRangeQueryForInline(shardingRuleConfig);
         Map<String, DataSource> dataSourceMap = new YamlDataSourceConfigurationSwapper().swapToDataSources(rootConfig.getDataSources());
         try {
-            return ShardingSphereDataSourceFactory.createDataSource(rootConfig.getDatabaseName(), dataSourceMap,
-                    Collections.singletonList(shardingRuleConfig), null);
+            return ShardingSphereDataSourceFactory.createDataSource(rootConfig.getDatabaseName(), dataSourceMap, Collections.singletonList(shardingRuleConfig), null);
             // CHECKSTYLE:OFF
         } catch (final Exception ex) {
             // CHECKSTYLE:ON

--- a/shardingsphere-jdbc/shardingsphere-jdbc-core/src/main/java/org/apache/shardingsphere/driver/data/pipeline/datasource/creator/ShardingSpherePipelineDataSourceCreator.java
+++ b/shardingsphere-jdbc/shardingsphere-jdbc-core/src/main/java/org/apache/shardingsphere/driver/data/pipeline/datasource/creator/ShardingSpherePipelineDataSourceCreator.java
@@ -47,7 +47,9 @@ public final class ShardingSpherePipelineDataSourceCreator implements PipelineDa
         try {
             return ShardingSphereDataSourceFactory.createDataSource(rootConfig.getDatabaseName(), dataSourceMap,
                     Collections.singletonList(shardingRuleConfig), null);
-        } catch (final SQLException ex) {
+            // CHECKSTYLE:OFF
+        } catch (final Exception ex) {
+            // CHECKSTYLE:ON
             closeDataSources(dataSourceMap.values());
             throw ex;
         }

--- a/shardingsphere-jdbc/shardingsphere-jdbc-core/src/main/java/org/apache/shardingsphere/driver/data/pipeline/datasource/creator/ShardingSpherePipelineDataSourceCreator.java
+++ b/shardingsphere-jdbc/shardingsphere-jdbc-core/src/main/java/org/apache/shardingsphere/driver/data/pipeline/datasource/creator/ShardingSpherePipelineDataSourceCreator.java
@@ -17,10 +17,11 @@
 
 package org.apache.shardingsphere.driver.data.pipeline.datasource.creator;
 
-import org.apache.shardingsphere.driver.api.ShardingSphereDataSourceFactory;
-import org.apache.shardingsphere.infra.config.algorithm.ShardingSphereAlgorithmConfiguration;
 import org.apache.shardingsphere.data.pipeline.api.datasource.config.impl.ShardingSpherePipelineDataSourceConfiguration;
 import org.apache.shardingsphere.data.pipeline.core.datasource.creator.PipelineDataSourceCreator;
+import org.apache.shardingsphere.driver.api.ShardingSphereDataSourceFactory;
+import org.apache.shardingsphere.infra.config.algorithm.ShardingSphereAlgorithmConfiguration;
+import org.apache.shardingsphere.infra.datasource.pool.destroyer.DataSourcePoolDestroyer;
 import org.apache.shardingsphere.infra.yaml.config.pojo.YamlRootConfiguration;
 import org.apache.shardingsphere.infra.yaml.config.swapper.YamlDataSourceConfigurationSwapper;
 import org.apache.shardingsphere.sharding.api.config.ShardingRuleConfiguration;
@@ -28,7 +29,9 @@ import org.apache.shardingsphere.sharding.yaml.swapper.ShardingRuleConfiguration
 
 import javax.sql.DataSource;
 import java.sql.SQLException;
+import java.util.Collection;
 import java.util.Collections;
+import java.util.Map;
 
 /**
  * ShardingSphere pipeline data source creator.
@@ -40,8 +43,14 @@ public final class ShardingSpherePipelineDataSourceCreator implements PipelineDa
         YamlRootConfiguration rootConfig = (YamlRootConfiguration) pipelineDataSourceConfig;
         ShardingRuleConfiguration shardingRuleConfig = ShardingRuleConfigurationConverter.findAndConvertShardingRuleConfiguration(rootConfig.getRules());
         enableRangeQueryForInline(shardingRuleConfig);
-        return ShardingSphereDataSourceFactory.createDataSource(rootConfig.getDatabaseName(), new YamlDataSourceConfigurationSwapper().swapToDataSources(rootConfig.getDataSources()),
-                Collections.singletonList(shardingRuleConfig), null);
+        Map<String, DataSource> dataSourceMap = new YamlDataSourceConfigurationSwapper().swapToDataSources(rootConfig.getDataSources());
+        try {
+            return ShardingSphereDataSourceFactory.createDataSource(rootConfig.getDatabaseName(), dataSourceMap,
+                    Collections.singletonList(shardingRuleConfig), null);
+        } catch (final SQLException ex) {
+            closeDataSources(dataSourceMap.values());
+            throw ex;
+        }
     }
     
     private void enableRangeQueryForInline(final ShardingRuleConfiguration shardingRuleConfig) {
@@ -50,6 +59,10 @@ public final class ShardingSpherePipelineDataSourceCreator implements PipelineDa
                 each.getProps().put("allow-range-query-with-inline-sharding", Boolean.TRUE.toString());
             }
         }
+    }
+    
+    private void closeDataSources(final Collection<DataSource> dataSourceMap) {
+        dataSourceMap.stream().map(DataSourcePoolDestroyer::new).forEach(DataSourcePoolDestroyer::asyncDestroy);
     }
     
     @Override

